common/elasticsearch/interfaces.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package elasticsearch
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/uber/cadence/common/metrics"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/log"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/service/config"
+)
+
+// NewGenericElasticSearchClient create a ES client
+func NewGenericElasticSearchClient(
+	connectConfig *config.ElasticSearchConfig,
+	visibilityConfig *config.VisibilityConfig,
+	logger log.Logger,
+) (GenericElasticSearch, error) {
+	// TODO hardcoded to V6 for now
+	return newV6Client(connectConfig, visibilityConfig, logger)
+}
+
+type (
+	// GenericElasticSearch is a generic interface for all versions of ElasticSearch clients
+	GenericElasticSearch interface {
+		Search(ctx context.Context, request *SearchRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		SearchByQuery(ctx context.Context, request *SearchByQueryRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		ScanByQuery(ctx context.Context, request *ScanByQueryRequest) (*p.InternalListWorkflowExecutionsResponse, error)
+		CountByQuery(ctx context.Context, index, query string) (int64, error)
+		PutMapping(ctx context.Context, index, root, key, valueType string) error
+		CreateIndex(ctx context.Context, index string) error
+		GetClosedWorkflowExecution(ctx context.Context, index string, request *p.InternalGetClosedWorkflowExecutionRequest) (*p.InternalGetClosedWorkflowExecutionResponse, error)
```

Comment:
Good point. I initially wanted to define different struct but at the end realize that it will be just the same as the persistence one, and to avoid having lots of code copying struct I simply use the same struct. 
I think a better idea is to use type alias -- this is the way we were using between persitenceInterfaces and dataInterfaces when some structs are exactly the same. This will make things more clear. And when it comes the time that one struct need to be different, we could separate them .